### PR TITLE
Add floating-point support

### PR DESCRIPTION
**Pull request by sgherbst**
## Summary
This PR adds support for synthesizable floating-point numbers via Berkeley HardFloat.  **svreal** still uses a fixed-point format by default, but this mode could be useful for comparison purposes.  To switch to the floating-point representation, the user just has to define the **HARD_FLOAT** flag.  More information (including where to find HardFloat and how to set up simulation / synthesis tools to use it) is given in the updated README.md.

## Details
* In general, code modifications are not needed to use **HARD_FLOAT**.  The main exceptions are **SYNC_ROM** and **SYNC_RAM**:
  * **SYNC_ROM**: the lookup table contents need to be updated to use HardFloat's **recfn** format.
  * **SYNC_RAM**: the width of the data signal may be different in the **HARD_FLOAT** case (since floating- and fixed-point formats do not necessarily have the same widths), so the user would need to account for that.
* Range, width, and exponent information is generally ignored in the floating-point case, since **HARD_FLOAT** uses a single floating-point format throughout the design.
* Updated tests so that they all run with fixed-point, **FLOAT_REAL**, and **HARD_FLOAT**.  As part of this, I refactored some common tester code into the **SvrealTester** class.
* More details are added as code comments in this PR.

### CODE REVIEW

File: .github/workflows/regression.yml [hunk 1]
```diff
@@ -43,7 +43,7 @@ jobs:
       run: python -m pip install --upgrade pip
     - name: Install dependencies
       run: |
-        brew install icarus-verilog
+        brew install icarus-verilog wget coreutils
```
Comment (sgherbst) [inline .github/workflows/regression.yml:46]: Added so that HardFloat can be installed with wget on macOS in the CI infrastructure

File: README.md [hunk 2]
```diff
@@ -1,19 +1,37 @@
 # svreal
 [![Actions Status](https://github.com/sgherbst/svreal/workflows/Regression/badge.svg)](https://github.com/sgherbst/svreal/actions)
-[![BuildKite Status](https://badge.buildkite.com/45ed712ae8720e9a3e7c040d2e3bc441a18b6ef269e4573723.svg)](https://buildkite.com/stanford-aha/svreal)
+[![BuildKite Status](https://badge.buildkite.com/45ed712ae8720e9a3e7c040d2e3bc441a18b6ef269e4573723.svg?branch=master)](https://buildkite.com/stanford-aha/svreal)
```
Comment (sgherbst) [inline README.md:3]: before this was targeting the latest BuildKite test, not necessarily on the master branch, which was problematic since it meant the front page might show a failing status even if the master branch was passing

File: codecov.yml [hunk 3]
```diff
@@ -8,7 +8,7 @@ coverage:
     project: # settings affecting project coverage
       default:
         target: auto # auto % coverage target
-        threshold: 5%  # allow for 5% reduction of coverage without failing
+        threshold: 10%  # allow for 10% reduction of coverage without failing
```
Comment (sgherbst) [inline codecov.yml:11]: There was so little Python code in the repo up until this point that we were at nearly 100% before.  I think we can bump this threshold back to 5% soon, now that we have a more reasonable amount of code for the coverage tool to check.

File: svreal/__init__.py [hunk 4]
```diff
@@ -1,6 +1,54 @@
+import os
 from pathlib import Path
 
+from svreal.constants import *
+from svreal.enums import RealType
+from svreal.hard_float import real2recfn, recfn2real
+from svreal.fixed_point import real2fixed, fixed2real, calc_fixed_exp
```
Comment (sgherbst) [inline svreal/__init__.py:7]: This file now uses the conversion functions provided by **svreal**, rather than re-implementing them here

File: svreal/svreal.sv [hunk 5]
```diff
@@ -46,6 +72,142 @@ endfunction
...
+
+    // assign the output
+    dbl_bits = {dbl_sign, dbl_exp, dbl_sig};
+    recfn2real = $bitstoreal(dbl_bits);
+endfunction
+
+`define REC_FN_TO_REAL(value) recfn2real(value)
+
+// convert a real number to the HardFloat recoded format
+function logic [((`HARD_FLOAT_EXP_WIDTH)+(`HARD_FLOAT_SIG_WIDTH)):0] real2recfn(input real in);
```
Comment (sgherbst) [inline svreal/svreal.sv:140]: this function is synthesizable (at least for parameter calculation) and is used to compute the representation of constants when HARD_FLOAT is defined

File: svreal/svreal.sv [hunk 6]
```diff
@@ -934,6 +1236,8 @@ module sync_ram_real #(
     // even when FLOAT_REAL is defined.
     `ifdef FLOAT_REAL
         assign out = `FIXED_TO_FLOAT(data, data_expt);
+    `elsif HARD_FLOAT
+        assign out = data;
```
Comment (sgherbst) [inline svreal/svreal.sv:1240]: note that when HARD_FLOAT is defined, the RAM contents are assumed to be in **recfn** format.  otherwise we would need extra hardware to convert the fixed-point RAM contents to **recfn**.

File: tests/common.py [hunk 7]
```diff
@@ -43,22 +56,39 @@ def pytest_synth_params(metafunc):
...
     # set defaults
     if src_files is None:
         src_files = []
     if hdr_files is None:
         hdr_files = []
+    if defines is None:
+        defines = {}
+
+    # add to source files
+    if real_type == RealType.HardFloat:
```
Comment (sgherbst) [inline tests/common.py:70]: we would probably want to do something similar in anasymod at some point, so that the HardFloat sources and headers are automatically pulled in if needed (but only if necessary, because we don't want HardFloat to be required)

File: tests/test_float.sv [hunk 8]
```diff
@@ -0,0 +1,121 @@
+`timescale 1ns/1ps
```
Comment (sgherbst) [inline tests/test_float.sv:1]: this new test checks the behavior of some corner-cases that are unique to floating point (infinities, nans, subnormal numbers, etc.).  One of the particularly tricky aspects here is that a normal number in the Python **float** type could be an infinity or subnormal in the **recfn** format, depending on the exponent and significand width.

File: tests/test_repr.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
+import pytest
```
Comment (sgherbst) [inline tests/test_repr.py:1]: this test checks the Python conversion functions used to convert real numbers to their binary representation.  most of these functions are implemented in svreal.sv, but it's useful to have a Python version of the function too

File: tests/common.py (hunk 7 above)
Comment (gabrielrutsch) [reply tests/common.py:70]: Sure makes sense. I think this can be easily managed, if we expose hard_float as an argument/option similarly to --float.

**Review by gabrielrutsch — APPROVED** [review]
